test/custom-xml-device-type.test.js

```diff
@@ -0,0 +1,274 @@
+    '{ 0x10E0FCA7, ZAP_ATTRIBUTE_INDEX(0), 0, 0, ZAP_CLUSTER_MASK(CLIENT), NULL }, /* Endpoint: 1, Cluster: Custom Cluster (client) */ \\'
+  )
+  expect(zapConfig).toContain(
+    '{ 0xABCDFFCD, ZAP_ATTRIBUTE_INDEX(0), 0, 0, ZAP_CLUSTER_MASK(CLIENT), NULL }, /* Endpoint: 1, Cluster: Test Cluster - Device Type (client) */ \\'
+  )
+
+  let zapConfigVer2 = genResult.content['zap-config-version-2.h']
+  expect(zapConfigVer2).toContain(
+    '{ 0x0302, ZCL_INT8U_ATTRIBUTE_TYPE, 1, (ATTRIBUTE_MASK_MANUFACTURER_SPECIFIC| ATTRIBUTE_MASK_CLIENT), { (uint8_t*)0  } }, /* 2 Cluster: Custom Cluster, Attribute: A9, Side: client*/ \\'
+  )
+  expect(zapConfigVer2).toContain(
+    '{ 0x0006, 0x02, COMMAND_MASK_OUTGOING_CLIENT }, /* 2, Cluster: On/off, Command: Toggle*/ \\'
+  )
+  expect(zapConfigVer2).toContain(
+    '{ 0xFFCD, 0x00, COMMAND_MASK_OUTGOING_CLIENT | COMMAND_MASK_MANUFACTURER_SPECIFIC }, /* 15, Cluster: Test Cluster - Device Type, Command: CommandOne*/ \\'
+  )
+})
+
+test('Load same custom xml with different zcl file', async () => {
+  await zclLoader.loadZcl(db, env.builtinMatterZclMetafile())
+  let uuid = util.createUuid()
+  let newSid = await testQuery.createSession(
+    db,
+    'USER',
+    uuid,
+    env.builtinMatterZclMetafile()
+  )
+
+  // load custom xml with device type
+  result = await zclLoader.loadIndividualFile(
+    db,
+    testUtil.testCustomXmlDeviceType,
+    newSid
+  )
+  expect(result.succeeded).toBe(true)
+
+  // Check if new device type cluster is correctly added
+  // There should be a new device type cluster for on/off with clusterRef from the matter zcl file
+  let deviceTypeClusters =
+    await queryDeviceType.selectDeviceTypeClustersByDeviceTypeRef(
+      db,
+      customDeviceType.id
+    )
+
+  expect(deviceTypeClusters.length).toBe(5)
+})
+
+test(`Load .zap file with custom xml with Device Type`, async () => {
+  let newSid = await querySession.createBlankSession(db)
+
+  // importing a zap file that has an endpoint with custom xml device type
+  let importResult = await importJs.importDataFromFile(
+    db,
+    testUtil.testMatterCustomZap2,
+    {
+      sessionId: newSid
+    }
+  )
+
+  expect(importResult.templateIds.length).toBe(1)
+
+  let genResult = await genEngine.generate(
+    db,
+    newSid,
+    importResult.templateIds[0],
+    {},
+    {
+      generateOnly: ['endpoint-config.c', 'endpoints.out'],
+      disableDeprecationWarnings: true
+    }
+  )
+  expect(genResult.hasErrors).toBeFalsy()
+
+  // Check if the generated files contain the expected content
+  let endpoints = genResult.content['endpoints.out']
+  expect(endpoints).toContain('  >> device: DUT-Server [2]')
+
+  let endpointConfig = genResult.content['endpoint-config.c']
+  expect(endpointConfig).toContain(
+    '{ 0x00000000, ZAP_TYPE(BOOLEAN), 1, 0, ZAP_SIMPLE_DEFAULT(0) }, /* OnOff */  \\'
+  )
+  expect(endpointConfig).toContain(
+    '/* Endpoint: 1, Cluster: Test Cluster - Device Type (server) */ \\'
+  )
+})
```

Comment:
Recently figured out how to open a file in the browser when zap server is running

http://localhost:9070/?standalone=false&filePath=%2abc%2Fconfig%2Fzcl%2Fzcl_config.zap#/

You could try using this in the cypress tests to open a .zap file with custom xml and check UI elements if you are curious.